[Analysis] Teach ScalarEvolution::getRangeRef about more dereferenceable objects

david-arm opened the pull request:
Whilst dealing with review comments on

https://github.com/llvm/llvm-project/pull/96752

I discovered that SCEV does not know about the dereferenceable attribute
on function arguments so I have updated getRangeRef to make use of it
by calling getPointerDereferenceableBytes.

david-arm committed:
[Analysis] Teach ScalarEvolution::getRangeRef about more dereferenceable objects

Whilst dealing with review comments on

https://github.com/llvm/llvm-project/pull/96752

I discovered that SCEV does not know about the dereferenceable
attribute on function arguments so I have updated getRangeRef
to make use of it by calling getPointerDereferenceableBytes.

nikic commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -6858,10 +6858,9 @@ const ConstantRange &ScalarEvolution::getRangeRef(
       ObjectSizeOpts Opts;
       Opts.RoundToAlign = false;
       Opts.NullIsUnknownSize = true;
```
Opts is unused now.

david-arm replied:
Done

nikic commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -6858,10 +6858,9 @@ const ConstantRange &ScalarEvolution::getRangeRef(
-           isAllocationFn(V, &TLI)) &&
-          getObjectSize(V, ObjSize, DL, &TLI, Opts) && ObjSize > 1) {
+      bool CanBeNull, CanBeFreed;
+      uint64_t ObjSize = V->getPointerDereferenceableBytes(DL, CanBeNull, CanBeFreed);
```
ObjSize -> DerefBytes or something. It may not be the full object size.

david-arm replied:
Done

nikic commented on llvm/test/Analysis/ScalarEvolution/no-wrap-add-exprs.ll:
```diff
@@ -362,3 +362,20 @@ loop:
 }
+
+
+define void @dereferenceable_arg(ptr dereferenceable(128) %len_addr, ptr dereferenceable(128) align(8) %len_addr2) {
```
Can you please also add a test using dereferenceable_or_null?

david-arm replied:
Done

nikic reviewed (commented):
I think this is correct. The logic does not really depend on it being the start of an allocated object.

david-arm committed:
Address review comments

* Fix code formatting issues.
* Remove unused ObjectSizeOpts variable.
* Renamed ObjSize -> DerefBytes.
* Added test for dereferenceable_or_null.
* While trying to make use of the CanBeNull variable that is set
by getPointerDereferenceableBytes I discovered a bug when the
Value is an alloca in a non-null address space. I fixed this by
making the behaviour consistent with llvm::isKnownNonZero.

david-arm commented:
Thanks for the quick review @nikic! I've tried to address your comments. I also discovered what seems to be a bug in getPointerDereferenceableBytes that I've fixed.

fhahn commented on llvm/test/Analysis/ScalarEvolution/no-wrap-add-exprs.ll:
```diff
@@ -362,3 +362,36 @@ loop:
 }
+
+
+define void @dereferenceable_arg(ptr dereferenceable(128) %len_addr, ptr dereferenceable(128) align(8) %len_addr2) {
```
might be worth to also add another pointer with a different size

fhahn reviewed (approved):
LGTM, thanks! Would be good to wait a bit in case @nikic has additional thoughts

nikic commented on llvm/lib/IR/Value.cpp:
```diff
@@ -909,7 +909,7 @@ uint64_t Value::getPointerDereferenceableBytes(const DataLayout &DL,
       DerefBytes =
           DL.getTypeStoreSize(AI->getAllocatedType()).getKnownMinValue();
-      CanBeNull = false;
+      CanBeNull = AI->getType()->getPointerAddressSpace() != 0;
```
I don't think this PR should be part of this change. Note that there is also a similar case for globals below.

I'm also not entirely sure this change should be made at all. There is a subtle distinction in semantics between:

 * Can return an allocation or null. This means that the pointer is *not* always dereferenceable.
 * Can return an allocation that may also be located at the null address. This means that the pointer *is* always dereferenceable, even if it is null.

The doc comment for this method is phrased vaguely enough that it's not super clear what the intended semantics are:

```
  /// If CanBeNull is set by this function the pointer can either be null or be
  /// dereferenceable up to the returned number of bytes.
```

But there are clearly some uses of the API where the current implementation is fine, for example
https://github.com/llvm/llvm-project/blob/9d364286f3b63e99ed3838f179aa2223f930f1ab/llvm/lib/Transforms/InstCombine/InstructionCombining.cpp#L3060 doesn't care whether the pointer is null, as long as it is dereferenceable.

The use here *does* care. I think what I would go for is to leave this method alone and drop the `!CanBeNull ||` check (maybe with a comment on why we don't use it).

david-arm replied:
Fair enough and thanks for explaining the subtle distinction! I've reverted the changes.

david-arm committed:
Revert some changes and add new test

* Reverted the changes to llvm/lib/IR/Value.cpp.
* Added another test in Analysis/ScalarEvolution/no-wrap-add-exprs.ll
for a differently sized pointer argument.

nikic reviewed (approved):
LGTM

fhahn commented:
> @fhahn @nikic Unless there are any objections I plan to land this patch later today as I think I've now addressed all concerns. Thanks!

Looks like the latest pre-commit test build has `Transforms/PhaseOrdering/scev-custom-dl.ll` failing. Could you double check if that's related to the patch?

david-arm committed:
Fix test scev-custom-dl.ll

david-arm commented:
> > @fhahn @nikic Unless there are any objections I plan to land this patch later today as I think I've now addressed all concerns. Thanks!
> 
> Looks like the latest pre-commit test build has `Transforms/PhaseOrdering/scev-custom-dl.ll` failing. Could you double check if that's related to the patch?

Yes it - good spot! I've fixed it now and run make check-all against HEAD of LLVM locally.
